### PR TITLE
Folder upload

### LINKED ISSUE #485: Folder Upload
### Expected behaviour
- The user clicks on '+ New' button and a new element is listed: 'Upload folder ...'
- A folder selection dialog opens - user choose a folder
- Upload starts upon hitting okay
- Walk the directory information and create folders and subfolders (including empty folders) and upload files
- Progress shall be visible in the upload progress component
- This option is only available on browsers which support folder upload https://caniuse.com/#feat=input-file-directory
- Rename exiting 'Upload' to 'Upload files'
- Choose a proper icon (maybe add a new icon to ODS)
- As an alternative way the user can drag and drop a folder from the desktop onto the browser

### PR DESCRIPTION
**Pull request by Julian1998**
## Description

Added folder upload

## Related Issue

- Fixes #485 

## How Has This Been Tested?

- press on new
- click on upload folder, select one
- check if all the folder (except empty folder got created) and the files got uploaded
- delete the folder
- reupload but this time use drag & drop
- alle folders and files should be created (also empty folders) correctly

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [ ] Tests

## Checklist:

- [x] Code changes
- [ ] Unit tests added
- [ ] Acceptance tests added
- [ ] Documentation ticket raised: <link> 

## Open tasks:

- [ ] write acceptance tests
- [x] fix PROPFIND Request

### CODE REVIEW

**Comment by ownclouders** [comment]
<details><summary>:boom: Acceptance tests <strong></strong> failed. Please find the screenshots inside ...</summary>

https://drone.owncloud.com/owncloud/phoenix/3533/24

<p>















</p></details>

**Comment by ownclouders** [comment]
<details><summary>:boom: Acceptance tests <strong>webUIUpload</strong> failed. Please find the screenshots inside ...</summary>

https://drone.owncloud.com/owncloud/phoenix/3533/19

<p>

























</p></details>

File: apps/files/src/mixins.js [hunk 1]
```diff
@@ -163,6 +163,48 @@ export default {
...
+          let folderName = folders[folder]
+          let exists = this.files.find((n) => {
+            if (n['name'] === folderName) {
+              return n
+            }
+          })
+          if (!exists) {
+            this.$client.files.createFolder(path + folderName)
+          }
+          this.navigateTo('files-list', path + folderName)
```
Comment (DeepDiver1975) [inline apps/files/src/mixins.js]: why nagivate to? I must be missing something ...

File: apps/files/src/mixins.js [hunk 2]
```diff
@@ -163,6 +163,48 @@ export default {
...
+        for (let folder in folders) {
+          console.log(this.files)
+          let folderName = folders[folder]
+          let exists = this.files.find((n) => {
+            if (n['name'] === folderName) {
+              return n
+            }
+          })
+          if (!exists) {
+            this.$client.files.createFolder(path + folderName)
```
Comment (DeepDiver1975) [inline apps/files/src/mixins.js]: any further processing (uploading content into the newly created folder) should be done async (.then())

**Comment by ownclouders** [comment]
<details><summary>:boom: Acceptance tests <strong>webUIUpload</strong> failed. Please find the screenshots inside ...</summary>

https://drone.owncloud.com/owncloud/phoenix/4109/19

<p>

























</p></details>

**Comment by ownclouders** [comment]
<details><summary>:boom: Acceptance tests <strong></strong> failed. Please find the screenshots inside ...</summary>

https://drone.owncloud.com/owncloud/phoenix/4109/24

<p>



























</p></details>

**Comment by ownclouders** [comment]
<details><summary>:boom: Acceptance tests <strong></strong> failed. Please find the screenshots inside ...</summary>

https://drone.owncloud.com/owncloud/phoenix/4109/23

<p>















</p></details>

File: apps/files/src/FileUpload.js [hunk 3]
```diff
@@ -1,21 +1,29 @@
...
 
-  upload (additionalData = {}) {
+  upload (options = {}) {
     let xhr = new XMLHttpRequest()
 
     // Headers
-    xhr.open(this.type, this.url + encodeURIComponent(this.file.name), true)
+    xhr.open(this.type, this.url + encodeURIComponent(this.path), true)
     xhr.responseType = 'text'
+    if (options.overwrite) {
```
Comment (DeepDiver1975) [inline apps/files/src/FileUpload.js]: looks like some rebase fuck up ....

File: apps/files/src/mixins.js [hunk 4]
```diff
@@ -144,38 +144,110 @@ export default {
...
+        this.$_ocUpload_addDirectoryToQue(e)
+      } else {
+        for (let item of items) {
+          item = item.webkitGetAsEntry()
+          if (item.isDirectory) {
+            if (!this.checkIfElementExists(item)) {
+              this.processDirectoryEntryRecursively(item)
+            } else {
+              this.showNotification({
+                title: this.$gettextInterpolate('Upload for %{folder} failed - Folder already exists', { folder: item.name }, true),
```
Comment (DeepDiver1975) [inline apps/files/src/mixins.js]: first translate the message and then interpolate ... see in the remaining code base ...

**Comment by DeepDiver1975** [comment]
I assume this is a browser warning - can we get rid of this? @Julian1998 

Screenshot from 2019-07-17 08-21-32

**Comment by DeepDiver1975** [comment]
There is some path building and encoding issue:
- [ ] too many slashes
- [ ] the other slash in the folder/file path is encoded

Screenshot from 2019-07-17 08-23-57

**Comment by DeepDiver1975** [comment]
- [ ] After uploading a folder (there have been errors) - the upload progress does not disapear
Screenshot from 2019-07-17 08-26-04

- [ ] after each upload a propfind is issued - this is not necessary (as discussed - I'll have a look)

**Comment by DeepDiver1975** [comment]
- [ ] higlight the uploaded folder - this will scroll to it - will work after #1264 is merged

File: apps/files/src/components/FilesAppBar.vue [hunk 5]
```diff
@@ -19,6 +19,7 @@
             <oc-drop toggle="#new-file-menu-btn" mode="click">
               <oc-nav>
                 <file-upload :url='url' :headers="headers" @success="onFileSuccess" @error="onFileError" @progress="onFileProgress"></file-upload>
+                <folder-upload :url='url' :headers="headers" @success="onFileSuccess" @error="onFileError" @progress="onFileProgress"></folder-upload>
```
Comment (DeepDiver1975) [inline apps/files/src/components/FilesAppBar.vue]: needs own success and error handler ....

File: apps/files/src/mixins.js [hunk 6]
```diff
@@ -164,29 +164,139 @@ export default {
...
+              })
+            }
+          })
         })
+      })
+    },
+    async $_ocUpload_addDropToQue (e) {
+      let items = e.dataTransfer.items || e.dataTransfer.files
 
+      // Fallback
```
Comment (DeepDiver1975) [inline apps/files/src/mixins.js]: fallback?

**Comment by ownclouders** [comment]
<details><summary>:boom: Acceptance tests <strong>webUISharingInternalUsers</strong> failed. Please find the screenshots inside ...</summary>

https://drone.owncloud.com/owncloud/phoenix/4155/9

<p>




</p></details>

**Comment by LukasHirt** [comment]
Can't upload any file at all now. No matter if I try drag and drop, file upload or folder upload. If I try to upload folder - the folder gets created but files still not.
image

**Comment by Julian1998** [comment]
> I assume this is a browser warning - can we get rid of this? @Julian1998
> 
> Screenshot from 2019-07-17 08-21-32

No you can't. It's a new security feature (https://stackoverflow.com/a/50473666)

**Comment by ownclouders** [comment]
<details><summary>:boom: Acceptance tests <strong>webUISharingInternalUsers</strong> failed. Please find the screenshots inside ...</summary>

https://drone.owncloud.com/owncloud/phoenix/4248/9

<p>




</p></details>

**Comment by Julian1998** [comment]
> There is some path building and encoding issue:
> 
> * [ ]  too many slashes
> * [ ]  the other slash in the folder/file path is encoded
> 
> Screenshot from 2019-07-17 08-23-57
> 
> unrelated to this PR -> new issue

Can't reproduce that one. I got two slahes (which is the correct behavior)

**Comment by ownclouders** [comment]
<details><summary>:boom: Acceptance tests <strong></strong> failed. Please find the screenshots inside ...</summary>

https://drone.owncloud.com/owncloud/phoenix/4248/13

<p>















</p></details>

**Comment by DeepDiver1975** [comment]
> Can't reproduce that one. I got two slahes (which is the correct behavior)

even two is one too many ;-)

**Pull request merged**